notebooks/samples/explanations/tf2/ai-explanations-image.ipynb

```diff
@@ -0,0 +1,1746 @@
+    "permutation = np.random.permutation(8 * 20)\n",
+    "some_flowers, some_labels = (some_flowers[permutation], some_labels[permutation])\n",
+    "\n",
+    "predictions = model.predict(some_flowers, batch_size=16)\n",
+    "evaluations = model.evaluate(some_flowers, some_labels, batch_size=16)\n",
+    "\n",
+    "print(np.array(CLASSES)[np.argmax(predictions, axis=-1)].tolist())\n",
+    "print('[val_loss, val_acc]', evaluations)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "colab": {},
+    "colab_type": "code",
+    "id": "qzCCDL1CZFx6"
+   },
+   "outputs": [],
+   "source": [
+    "display_9_images_with_predictions(some_flowers, predictions, some_labels)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "kV_NEAQwwH0e"
+   },
+   "source": [
+    "## Export the model as a TF 2.x SavedModel\n",
+    "\n",
+    "When using TensorFlow 2.x, you export the model as a `SavedModel` and load it into Cloud Storage. During export, you need to define a serving function to convert data to the format your model expects. If you send encoded data to AI Platform, your serving function ensures that the data is decoded on the model server before it is passed as input to your model.\n",
+    "\n",
+    "### Serving function for image data\n",
+    "\n",
+    "Sending base 64 encoded image data to AI Platform is more space efficient. Since this deployed model expects input data as raw bytes, you need to ensure that the b64 encoded data gets converted back to raw bytes before it is passed as input to the deployed model.\n",
+    "\n",
+    "To resolve this, define a serving function (`serving_fn`) and attach it to the model as a preprocessing step. Add a `@tf.function` decorator so the serving function is part of the model's graph (instead of upstream on a CPU).\n",
+    "\n",
+    "When you send a prediction or explanation request, the request goes to the serving function (`serving_fn`), which preprocesses the b64 bytes into raw numpy bytes (`preprocess_fn`). At this point, the data can be passed to the model (`m_call`)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "colab": {},
+    "colab_type": "code",
+    "id": "2CRJTE7BXPJE"
+   },
+   "outputs": [],
+   "source": [
+    "export_path = 'gs://' + BUCKET_NAME + '/explanations/mymodel'\n",
+    "\n",
+    "\n",
+    "def _preprocess(bytes_input):\n",
+    "    decoded = tf.io.decode_jpeg(bytes_input, channels=3)\n",
+    "    decoded = tf.image.convert_image_dtype(decoded, tf.float32)\n",
+    "    resized = tf.image.resize(decoded, size=(192, 192))\n",
+    "    return resized\n",
+    "\n",
+    "\n",
+    "@tf.function(input_signature=[tf.TensorSpec([None], tf.string)])\n",
+    "def preprocess_fn(bytes_inputs):\n",
+    "    decoded_images = tf.map_fn(_preprocess, bytes_inputs, dtype=tf.float32, back_prop=False)\n",
```

Comment:
Let's switch this line to
with tf.device("cpu:0"): 
        decoded_images = tf.map_fn(_preprocess, bytes_inputs, dtype=tf.float32)

so that this notebook runs on gpus. It's an issue with tensorflow and should be fixed in 2.3.

https://github.com/tensorflow/tensorflow/issues/28007\n"

Reply:
Making sure I get the indentation right... Is this what you mean?

```
@tf.function(input_signature=[tf.TensorSpec([None], tf.string)])
def preprocess_fn(bytes_inputs):
    with tf.device("cpu:0"):
        decoded_images = tf.map_fn(_preprocess, bytes_inputs, dtype=tf.float32, back_prop=False)
    return {"numpy_inputs": decoded_images}  # User needs to make sure the key matches model's input
```

Reply:
correct. Removing back_prop will make it work and reduce warnings.

Reply:
Oops, so:

```
@tf.function(input_signature=[tf.TensorSpec([None], tf.string)])
def preprocess_fn(bytes_inputs):
    with tf.device("cpu:0"):
        decoded_images = tf.map_fn(_preprocess, bytes_inputs, dtype=tf.float32)
    return {"numpy_inputs": decoded_images}  # User needs to make sure the key matches model's input
```

Reply:
cpu:0 means that the static graph will run upstream from GPU on the CPU?

Reply:
Yes. cpu:0 means this op will be executed on cpus.

Reply:
https://www.tensorflow.org/guide/gpu#manual_device_placement

Reply:
Added. Thanks!